src/pyref_map.rs

```diff
@@ -0,0 +1,194 @@
+
+use std::ptr::NonNull;
+use std::marker::PhantomData;
+use std::ops::{Deref, DerefMut};
+
+use crate::pyclass::PyClass;
+use crate::pycell::{PyRef, PyRefMut};
+use crate::pyclass::boolean_struct::{True, False, private::Boolean};
+
+
+/// Represents a `PyRef` or `PyRefMut` with an opaque pyclass type.
+trait OpaquePyRef<'py>: 'py {}
+
+impl<'py, T: PyClass> OpaquePyRef<'py> for PyRef<'py, T> {}
+impl<'py, T: PyClass<Frozen=False>> OpaquePyRef<'py> for PyRefMut<'py, T> {}
+
+
+/// Base wrapper type for a [`PyRef<'py, T>`] or [`PyRefMut<'py, T>`] that 
+/// dereferences to data of type `U` that is nested within a pyclass `T` 
+/// instead of `T` itself. 
+/// 
+/// See the type aliases [`PyRefMap<'py, U>`] and [`PyRefMapMut<'py, U>`] 
+/// for more information.
+pub struct PyRefMapBase<'py, U: 'py, Mut: Boolean> {
+    // Either `PyRef` or `PyRefMut` guarding the opaque pyclass from which 
+    // the target is borrowed. 
+    owner: Box<dyn OpaquePyRef<'py>>,
+    // Pointer to the `Deref` target which is (probably) borrowed from `owner`.
+    // This pointer is derived from either `&U` or `&mut U` as  indicated by 
+    // the `Mut` parameter.
+    target: NonNull<U>,
+    // Marks whether mutable methods are supported. If `Mut` is `True` then 
+    // `owner` is a `PyRefMut` and `target` was derived from `&mut U`, so 
+    // the pointer may be mutably dereferenced safely; if `False`, then 
+    // `owner` may be either `PyRef` or `PyRefMut` and `target` was derived 
+    // from `&U` so mutable dereferencing is forbidden.
+    _mut: PhantomData<Mut>
+}
```

Comment:
Thanks for the suggestion, I'll play around with it a bit to see if it can be simplified. 

However, I'm not sure that your method would work unless I am missing something. My motivation for storing the original `PyRef`/`PyRefMut` within the `PyRefMap` struct was to hold the borrow on the `T: PyClass` that the `NonNull<U>` is pointing into so that the pointer is guaranteed to be safely dereferenceable.  If I allow the `PyRef` to be dropped, its `Drop` impl will call `release_borrow` and stop protecting `T` from being mutably aliased. The inner `Bound<T>` would also get dropped, which would decrement the reference count and potentially let `T` get garbage collected. The next time `PyRefMap` get's dereferenced would then be UB. 

So to prevent all this I would have to do something messy, perhaps `ManuallyDrop` the `PyRef` and then add a `Drop` impl for `PyRefMap` that releases the borrow; and then I'd still have to find a way to decrement the `T`'s reference count, which wouldn't be possible if I only had access to `&dyn PyClassBorrowChecker`. I'm also not sure how I would even get a reference to the borrow checker with the `'py` lifetime; it seems like the only way to get it is by borrowing it from a `&'a Bound<'py, T>` which could only give you `&'a dyn PyClassBorrowChecker`. But even if there is a way to get around all this, it seems a lot more complicated/unsafe to me than just storing the original `PyRef` and letting it handle all the protection and cleanup with its existing mechanisms.

Reply:
>  If I allow the PyRef to be dropped, its Drop impl will call release_borrow and stop protecting...

Of course. I was just sketching out an approach. I would really prefer to not allocate in `PyRefMap::new`, this is another approach to erase the pyclass type without boxing things.

A solution would be to refactor that into its own struct:
```rust
pub struct BorrowRef<'b>{
     borrow:  &'b ???,
}
impl Drop for BorrowRef {
    ...
}

pub struct PyRef<'p, T: PyClass> {
    inner: Bound<'p, T>,
    borrow: BorrowRef<'p>
}

pub struct PyRefMap<'b, U>{
     value: NonNull<U>
     borrow: BorrowRef<'p>
}
```

Another solution is your simplified solution that stores the PyRef inline but does not erase the type.